### PR TITLE
Evict patient from experiment if their visit facility is deleted

### PR DESCRIPTION
**Pull request by qptr**
An error occurs while recording a visit for a patient enrolled in an experiment (also called a membership), if their visit facility has been soft deleted.

Since this is invalid data that can affect the experiment's outcomes, and also disrupts the daily experiment runner, these memberships will be evicted from the experiment.

Another change this PR makes is to filter out evicted memberships that have visited but haven't been marked as visited, from those whose visits need to be marked. This is for 2 reasons:
1. Memberships who were evicted due to their visit facilities being soft deleted won't get picked up again to record the invalid visit
2. Evicted patients aren't expected to continue to be part of the experiment flow

### CODE REVIEW

File: spec/models/experimentation/treatment_group_membership_spec.rb [hunk 1]
```diff
@@ -83,6 +83,22 @@
...
+      visit_facility = create(:facility)
+      visit = create(:prescription_drug, device_created_at: 2.days.ago, patient: membership.patient, facility: visit_facility)
+
+      visit_facility.discard!
+      visit.reload
+
+      membership.record_visit(blood_pressure: nil, blood_sugar: nil, prescription_drug: visit)
+      membership.reload
+
+      expect(membership.status).to eq("evicted")
```
Comment (danySam) [inline spec/models/experimentation/treatment_group_membership_spec.rb:98]: Do you want to add a negative test for non "evicted" scenarios?

**Review by danySam — APPROVED** [review]

**Review by PriyangaPKini — APPROVED** [review]

File: app/models/experimentation/treatment_group_membership.rb [hunk 2]
```diff
@@ -22,18 +22,24 @@ def self.evict(reason:)
     def record_notification(reminder_template_id, notification)
       reload # this is to reload the `messages` field to avoid staleness while updating.
       self.messages ||= {}
+      localized_message = notifiaction_localized_message(notification)
```
Comment (govindkrjoshi) [inline app/models/experimentation/treatment_group_membership.rb:25]:
```ruby
localized_message = notifiaction_localized_message(notification) || ""
```

File: app/models/experimentation/treatment_group_membership.rb [hunk 3]
```diff
@@ -22,18 +22,24 @@ def self.evict(reason:)
...
         created_at: notification.created_at.to_s
       }
       save!
+    end
+
+    def notifiaction_localized_message(notification)
+      notification.localized_message
     rescue Messaging::MissingReferenceError => error
       Rails.logger.error("Failed fetching localised message on notification #{notification.id}, reason: #{error.message}")
+      ""
```
Comment (govindkrjoshi) [inline app/models/experimentation/treatment_group_membership.rb:42]: With the above change this is not necessary

Comment (qptr) [reply app/models/experimentation/treatment_group_membership.rb:42]: Rails.logger.error returns true @govindkrjoshi

**Pull request merged**